MoreLinq/TrySingle.cs

```diff
@@ -0,0 +1,171 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Attempts to retrieve and project the only element of a sequence,
+        /// similar to <see cref="Enumerable.Single{T}(IEnumerable{T})"/>.
+        /// Unlike that extension method if the sequence contains zero or many
+        /// elements this operator doesn't throw an exception, rather it provides
+        /// the <paramref name="resultSelector"/> function with the observed
+        /// cardinality. This enables the caller to differentiate between the
+        /// case where a sequence contains zero elements and many elements.
+        /// </summary>
+        /// <param name="source">
+        /// The source sequence that will be tested for its cardinality.</param>
+        /// <param name="zero">
+        /// The value that is passed as the first argument to
+        /// <paramref name="resultSelector" /> if the sequence has zero
+        /// elements.</param>
+        /// <param name="one">
+        /// The value that is passed as the first argument to
+        /// <paramref name="resultSelector" /> if the sequence has a
+        /// single element only.</param>
+        /// <param name="many">
+        /// The value that is passed as the first argument to
+        /// <paramref name="resultSelector" /> if the sequence has two or
+        /// more elements.</param>
+        /// <param name="resultSelector">
+        /// A function that is provided with the cardinality, and if the
+        /// sequence has just one element, the value of that element. Then
+        /// transforms the result to an instance of TResult.</param>
+        /// <typeparam name="T">
+        /// The type of the elements of the sequence.</typeparam>
+        /// <typeparam name="TCardinality">
+        /// The type that expresses cardinality.</typeparam>
+        /// <typeparam name="TResult">
+        /// The result type of the <paramref name="resultSelector"/> function.
+        /// </typeparam>
+        /// <returns>
+        /// The value returned by the <paramref name="resultSelector"/>.
+        /// </returns>
+        /// <remarks>
+        /// This operator uses immediate execution, but never consumes more
+        /// than two elements from the sequence. When the source sequence is an
+        /// <see cref="IList{T}"/> or <see cref="ICollection{T}"/> then the
+        /// implementation optimizes by checking the number of elements in
+        /// the underlying sequence.</remarks>
+        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
```

Comment:
Add a blank line between doc summary and method declaration for clarity:

```suggestion

        public static TResult TrySingle<T, TCardinality, TResult>(this IEnumerable<T> source,
```